pkg/actions/lua/net/http/request.go

```diff
@@ -0,0 +1,133 @@
+package http
+
+import (
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/Shopify/go-lua"
+	"github.com/treeverse/lakefs/pkg/actions/lua/util"
+)
+
+const defaultRequestTimeout = 30 * time.Second
+
+func Open(l *lua.State) {
+	open := func(l *lua.State) int {
+		lua.NewLibrary(l, httpLibrary)
+		return 1
+	}
+	lua.Require(l, "net/http", open, false)
+	l.Pop(1)
+}
+
+var httpLibrary = []lua.RegistryFunction{
+	{Name: "request", Function: httpRequest},
+}
+
+// httpRequest - perform http request
+//
+//	Accepts arguments (url, body) or table with url, method, body, headers. Value for url is required.
+//		method is by default GET or POST in case body is set.
+//	Returns code, body, headers, status.
+func httpRequest(l *lua.State) int {
+	req := prepareRequest(l)
+	client := http.Client{
+		Timeout: defaultRequestTimeout,
+	}
+	resp, err := client.Do(req)
+	check(l, err)
+	defer func() { _ = resp.Body.Close() }()
+	body, err := io.ReadAll(resp.Body)
+	check(l, err)
+
+	// push return
+	l.PushInteger(resp.StatusCode)
+	l.PushString(string(body))
+	pushResponseHeader(l, resp.Header)
+	l.PushString(resp.Status)
+	return 4
+}
+
+func prepareRequest(l *lua.State) *http.Request {
+	var (
+		reqMethod  = http.MethodGet
+		reqURL     string
+		reqBody    io.Reader
+		reqHeaders map[string]interface{}
+	)
+	switch l.TypeOf(1) {
+	case lua.TypeString:
+		// url and optional body
+		reqURL = lua.CheckString(l, 1)
+		if s, ok := l.ToString(2); ok {
+			reqBody = strings.NewReader(s)
+			reqMethod = http.MethodPost
+		}
+	case lua.TypeTable:
+		// extract request parameters from table
+		value, err := util.PullTable(l, 1)
+		check(l, err)
+		tbl := value.(map[string]interface{})
+		if s, ok := tbl["url"].(string); ok {
+			reqURL = s
+		}
+		if s, ok := tbl["body"].(string); ok {
+			reqBody = strings.NewReader(s)
+			reqMethod = http.MethodPost
+		}
+		if s, ok := tbl["method"].(string); ok {
+			reqMethod = s
+		}
+		if m, ok := tbl["headers"].(map[string]interface{}); ok {
+			reqHeaders = m
+		}
+	default:
+		lua.Errorf(l, "first argument can be url or request table (invalid type: %d)", l.TypeOf(1))
+		panic("unreachable")
```

Comment:
Why is it unreachable?

Reply:
Same as the above - we raise panic for the go-lua engine to handle the error - by design.